021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,304 @@
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const serialisedCookies = JSON.stringify({
+  a_session_PROJECT_ID: sessionToken,
+});
+
+client.headers["X-Fallback-Cookies"] = serialisedCookies;
+
+const account = new Account(client);
+const currentUser = account.get();
+```
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+**Problem #1: Solution A - Return session token in response body**
+
+Return the token within the session object, resulting in the follow example code:
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const account = new Account(client);
+const session = await account.createAnonymousSession();
+const sessionToken = session.token;
+```
+
+**Problem #1: Solution B - Include headers within the response**
+
+Return the session cookie in the response headers.
+
+```js
+import { parse } from "set-cookie-parser";
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const account = new Account(client);
+const session = await account.createAnonymousSession();
+
+const sessionToken = parse(session.headers["set-cookie"]).find((cookie) =>
+  cookie.name.startsWith("a_session")
+).value;
+```
+
+**Problem #2: Solution A - Temporary OAuth2 token exchange for session**
+
+- Modify the oauth2 flow to include a temporary token in the final redirect.
+- Implement a new endpoint, or extend the existing functionality of the magic URL endpoint, to exchange the temporary token for a session token.
+
+![SSR OAuth2 Flow Sequence Diagram](ssr-oauth-flow.png)
+
+1. Browser makes a GET request to the SSR application.
+2. SSR application makes a GET request to the Appwrite API.
+3. Appwrite API redirects the SSR application to the OAuth2 provider.
+4. The SSR application redirects the client to the OAuth2 provider.
+5. User authenticates with the OAuth2 provider.
+6. OAuth2 provider redirects the browser back to the Appwrite API.
+7. Appwrite API sets the session cookie on the Appwrite domain.
+8. Appwrite API redirects the browser back to the client application. **Now, the redirect URL includes a userId & temporary token.** e.g. `myssrapp.com/oauth2/success?userId=loks0n&token=TEMPORARY_TOKEN`
+9. The SSR application makes a POST request to the Appwrite API to exchange the temporary token for a session token.
+
+The SSR application must set up the success page to call the exchange endpoint. The page can then set session cookie on the SSR domain.
+`
+
+**Problem #3: Solution A - setSession helper method**
+
+A new SDK helper method, called `setSession`, to set the session token of future requests.
```

Comment:
SetSession or SetToken? 🤔

Reply:
I feel we generally use token for temporary keys.
This helper could take either the entire session object, or just the secret, in which case `setSecret` or `setSessionSecret` might be worth considering

Reply:
We have `setKey` for setting API key. So to follow that we should do `setSecret` for setting session secret.

But that sounds so undescriptive.. I would vote for `setSessionSecret`, but then we should consider doing `setApiKey`.

Reply:
I could live with `setSecret